### PR TITLE
Add the Meta Chunker class

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/77988

This PR adds the `Tribe__Meta__Chunker` class to allow us to transparently register potentially big and DB killing meta values to avoid killing the DB itself.
Only caveat: no more than one chunked instance for each meta value;  as if any call is an update one.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | String &quot;_tribe_chunker_&quot; does not require double quotes; use single quotes instead | [Chunker.php:59](https://github.com/moderntribe/tribe-common/blob/4245f432ab2a1c25bb04dd5de305a2eb775e9981/src/Tribe/Meta/Chunker.php#L59)
:x: | Each line in an array declaration must end in a comma | [Chunker.php:260](https://github.com/moderntribe/tribe-common/blob/4245f432ab2a1c25bb04dd5de305a2eb775e9981/src/Tribe/Meta/Chunker.php#L260)
:x: | String &quot;(%d, %s, %s)&quot; does not require double quotes; use single quotes instead | [Chunker.php:399](https://github.com/moderntribe/tribe-common/blob/4245f432ab2a1c25bb04dd5de305a2eb775e9981/src/Tribe/Meta/Chunker.php#L399)
:x: | String &quot;/(\d+)&quot; does not require double quotes; use single quotes instead | [Chunker.php:505](https://github.com/moderntribe/tribe-common/blob/4245f432ab2a1c25bb04dd5de305a2eb775e9981/src/Tribe/Meta/Chunker.php#L505)
:x: | String &quot;(.*)/&quot; does not require double quotes; use single quotes instead | [Chunker.php:505](https://github.com/moderntribe/tribe-common/blob/4245f432ab2a1c25bb04dd5de305a2eb775e9981/src/Tribe/Meta/Chunker.php#L505)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | String &quot;_tribe_chunker_&quot; does not require double quotes; use single quotes instead | [Chunker.php:59](https://github.com/moderntribe/tribe-common/blob/aba481723fda502443563080a4ed4fd7478248fc/src/Tribe/Meta/Chunker.php#L59)
:x: | Each line in an array declaration must end in a comma | [Chunker.php:260](https://github.com/moderntribe/tribe-common/blob/aba481723fda502443563080a4ed4fd7478248fc/src/Tribe/Meta/Chunker.php#L260)
:x: | String &quot;(%d, %s, %s)&quot; does not require double quotes; use single quotes instead | [Chunker.php:396](https://github.com/moderntribe/tribe-common/blob/aba481723fda502443563080a4ed4fd7478248fc/src/Tribe/Meta/Chunker.php#L396)
:x: | String &quot;/(\d+)&quot; does not require double quotes; use single quotes instead | [Chunker.php:502](https://github.com/moderntribe/tribe-common/blob/aba481723fda502443563080a4ed4fd7478248fc/src/Tribe/Meta/Chunker.php#L502)
:x: | String &quot;(.*)/&quot; does not require double quotes; use single quotes instead | [Chunker.php:502](https://github.com/moderntribe/tribe-common/blob/aba481723fda502443563080a4ed4fd7478248fc/src/Tribe/Meta/Chunker.php#L502)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | String &quot;_tribe_chunker_&quot; does not require double quotes; use single quotes instead | [Chunker.php:60](https://github.com/moderntribe/tribe-common/blob/4dc3c127250a053d63467bbb52f2233487f51f1f/src/Tribe/Meta/Chunker.php#L60)
:x: | Each line in an array declaration must end in a comma | [Chunker.php:261](https://github.com/moderntribe/tribe-common/blob/4dc3c127250a053d63467bbb52f2233487f51f1f/src/Tribe/Meta/Chunker.php#L261)
:x: | String &quot;(%d, %s, %s)&quot; does not require double quotes; use single quotes instead | [Chunker.php:397](https://github.com/moderntribe/tribe-common/blob/4dc3c127250a053d63467bbb52f2233487f51f1f/src/Tribe/Meta/Chunker.php#L397)
:x: | String &quot;/(\d+)&quot; does not require double quotes; use single quotes instead | [Chunker.php:503](https://github.com/moderntribe/tribe-common/blob/4dc3c127250a053d63467bbb52f2233487f51f1f/src/Tribe/Meta/Chunker.php#L503)
:x: | String &quot;(.*)/&quot; does not require double quotes; use single quotes instead | [Chunker.php:503](https://github.com/moderntribe/tribe-common/blob/4dc3c127250a053d63467bbb52f2233487f51f1f/src/Tribe/Meta/Chunker.php#L503)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Meta/Chunker.php [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+
+	/**
+	 * @var int The filter priority at which Chunker will operate on meta CRUD operations.
+	 */
+	protected $filter_p = - 1;
+
+	/**
+	 * @var string The meta key prefix applied ot any Chunker related post meta.
+	 */
+	protected $meta_key_prefix = "_tribe_chunker_";
```
Comment (bordoni) [inline src/Tribe/Meta/Chunker.php]: Can we use `'`?

File: src/Tribe/Meta/Chunker.php [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+	 *
+	 * @param int    $object_id
+	 * @param string $meta_key
+	 */
+	protected function delete_chunks( $object_id, $meta_key ) {
+		/** @var wpdb $wpdb */
+		global $wpdb;
+		$criteria = array(
+			'post_id'  => $object_id,
+			'meta_key' => $this->get_chunk_meta_key( $meta_key )
```
Comment (bordoni) [inline src/Tribe/Meta/Chunker.php]: Comma

File: src/Tribe/Meta/Chunker.php [hunk 3]
```diff
@@ -0,0 +1,584 @@
+<?php
+
+/**
+ * Class Tribe__Meta__Chunker
+ *
+ * Chunks large meta to avoid kiling the database in queries.
```
Comment (barryhughes) [inline src/Tribe/Meta/Chunker.php]: *Killing

File: src/Tribe/Meta/Chunker.php [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+ *
+ * Databases can have a `max_allowed_packet` value set as low as 1M; we often need to store large blobs of data way
+ * over that and doing so would kill hte database ("MySQL server has gone away"...); registering a meta to be chunked
```
Comment (barryhughes) [inline src/Tribe/Meta/Chunker.php]: *the

**Comment by lucatume** [comment]
Ok, the Meta Chunker has  now support for all meta queries too, those made using code like this:

```php
$all_meta = get_post_meta( $post_id );
```

and ready for another pass.

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | String &quot;update_post_metadata&quot; does not require double quotes; use single quotes instead | [Chunker.php:89](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L89)
:x: | String &quot;delete_post_metadata&quot; does not require double quotes; use single quotes instead | [Chunker.php:90](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L90)
:x: | String &quot;add_post_metadata&quot; does not require double quotes; use single quotes instead | [Chunker.php:91](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L91)
:x: | String &quot;get_post_metadata&quot; does not require double quotes; use single quotes instead | [Chunker.php:92](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L92)
:x: | There must be a single space after the first parenthesis | [Chunker.php:285](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L285)
:x: | There must be a single space before the last parenthesis | [Chunker.php:285](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L285)
:x: | ! must be followed by a single space | [Chunker.php:285](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L285)
:warning: | The method parameter $single is never used | [Chunker.php:518](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L518)
:x: | Each line in an array declaration must end in a comma | [Chunker.php:693](https://github.com/moderntribe/tribe-common/blob/21024799a8dba53e5a13ae5a44f98ece92faadb7/src/Tribe/Meta/Chunker.php#L693)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Meta/Chunker.php [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+	 * @var string The meta key prefix applied ot any Chunker related post meta.
+	 */
+	protected $meta_key_prefix = '_tribe_chunker_';
+
+	/**
+	 * @var int The largest size allowed by the Chunker.
+	 */
+	protected $max_chunk_size;
+
+	public function __construct(  ) {
```
Comment (bordoni) [inline src/Tribe/Meta/Chunker.php]: Remove spaces inside `()`

File: src/Tribe/Meta/Chunker.php [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+	 *
+	 * @return array|int|null|object
+	 */
+	public function get_max_chunk_size() {
+		if ( ! empty( $this->max_chunk_size ) ) {
+			return $this->max_chunk_size;
+		}
+		/** @var wpdb $wpdb */
+		global $wpdb;
+		$max_size = $wpdb->get_results( "SHOW VARIABLES LIKE 'max_allowed_packet';", ARRAY_A );
```
Comment (bordoni) [inline src/Tribe/Meta/Chunker.php]: Why not use `get_var`

Comment (lucatume) [inline src/Tribe/Meta/Chunker.php]: this is not a var really and `get_var` will not work in this case: I've tried.

**Comment by bordoni** [comment]
👍

**Pull request merged**